### DISCUSSION TITLE
Probe Macro - tool setter / tool height

### DISCUSSION THREAD

**Discussion by mtheriaulr** [terjeio/ioSender#362] [General]
Is it possible to use a macro to run the tool length offset with the "probe fixture @ G59.3" and "Establish Reference Offset" selected?

Ideally in the same custom macro that probes z height. (where the tool is already positioned over the z probe position.)

In a nutshell, would be able to position the first tool over the position to do a z probe and start the macro which would establish Z height at that location and then move to G59.3 and establish the reference offset, all with one press of a macro.

If so, what would be the macro for this? (I have a macro to probe Z and establish Z0, I just don't' know how to initiate the tool length probe from a macro)

**Comment by terjeio** [comment]
I am not sure how this can be done, maybe something like this:

```
G59.3
to safe Z with G28? 
G0X0Y0
G0Z0
G38.2Z...
$TLR
to safe Z with G28? 
G54
...
```
I am wary of probing from macros since the macro continue execution regardless of the probing failed/ok result.

A more advanced option is to enable [expression support](https://github.com/grblHAL/core/wiki/Expressions-and-flow-control) and run the macro from a SD card or a flash based filing system if available. This will allow you to perform error checking, saving and returning the original position etc.

---

The code behind the _Tool length offset_ tab can be found [here](https://github.com/terjeio/ioSender/blob/master/CNC%20Controls%20Probing/CNC%20Controls%20Probing/ToolLengthControl.xaml.cs).
